pkg/virt-controller/services/renderresources.go

```diff
@@ -0,0 +1,554 @@
+package services
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	v1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/kubevirt/pkg/downwardmetrics"
+	"kubevirt.io/kubevirt/pkg/util"
+	"kubevirt.io/kubevirt/pkg/util/hardware"
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+)
+
+type ResourceRendererOption func(renderer *ResourceRenderer)
+
+type ResourceRenderer struct {
+	vmLimits           k8sv1.ResourceList
+	vmRequests         k8sv1.ResourceList
+	calculatedLimits   k8sv1.ResourceList
+	calculatedRequests k8sv1.ResourceList
+}
+
+type resourcePredicate func(*v1.VirtualMachineInstance) bool
+
+type VMIPredicates struct {
+	resourceRules []VMIResourceRule
+	vmi           *v1.VirtualMachineInstance
+}
+
+type VMIResourceRule struct {
+	predicate resourcePredicate
+	option    ResourceRendererOption
+}
+
+func not(p resourcePredicate) resourcePredicate {
+	return func(instance *v1.VirtualMachineInstance) bool {
+		return !p(instance)
+	}
+}
+func NewVMIResourceRule(p resourcePredicate, option ResourceRendererOption) VMIResourceRule {
+	return VMIResourceRule{predicate: p, option: option}
+}
+
+func NewVMIPredicates(vmi *v1.VirtualMachineInstance, clusterConfig *virtconfig.ClusterConfig, networkResources map[string]string) VMIPredicates {
+	memoryOverhead := GetMemoryOverhead(vmi, clusterConfig.GetClusterCPUArch())
+	return VMIPredicates{
+		vmi: vmi,
+		resourceRules: []VMIResourceRule{
+			NewVMIResourceRule(doesVMIRequireDedicatedCPU, WithCPUPinning(vmi.Spec.Domain.CPU)),
+			NewVMIResourceRule(not(doesVMIRequireDedicatedCPU), WithoutDedicatedCPU(vmi.Spec.Domain.CPU, clusterConfig.GetCPUAllocationRatio())),
+			NewVMIResourceRule(util.HasHugePages, WithHugePages(vmi.Spec.Domain.Memory, memoryOverhead)),
+			NewVMIResourceRule(not(util.HasHugePages), WithMemoryOverhead(vmi.Spec.Domain.Resources, memoryOverhead)),
+			NewVMIResourceRule(func(instance *v1.VirtualMachineInstance) bool {
+				return len(networkResources) > 0
+			}, WithNetworkResources(networkResources)),
+			NewVMIResourceRule(util.IsGPUVMI, WithGPUs(vmi.Spec.Domain.Devices.GPUs)),
+			NewVMIResourceRule(util.IsHostDevVMI, WithHostDevices(vmi.Spec.Domain.Devices.HostDevices)),
+			NewVMIResourceRule(util.IsSEVVMI, WithSEV()),
+		},
```

Comment:
This is a very interesting idea, I like it a lot!

Small note: some predicates here don't do much, like the SEV example (the predicate check if `vmi.Spec.Domain.LaunchSecurity.SEV` is nil, and the option sets it to not be nil). But I still think you made the right choice introducing these predicates, since in the future the situation might be different.

Reply:
I considered moving the code that accesses the network resources inside the predicate, but that would just lead to duplicate code - the net resources are also required for the env. variables. 

I think for now this is how it has to be.